### ISSUE TITLE
Views not respecting API Docs access control

### ISSUE THREAD

**Issue by kscheirer** [apigee/apigee-api-catalog-drupal#55]
**Describe the bug**
A user who doesn't have access to api doc entities will still see a full list of apidocs at the /apis view.

**To Reproduce**
1. Go to `admin/structure/views/view/apigee_api_catalog`
2. Set Page Settings Access to `Unrestricted`
3. Go to `admin/people/permissions`
4. Uncheck the Anonymous role for `View published API docs`, leaving just Authenticated checked
5. Flush all caches
6. Rebuild node access (link is available on status report page)

**Expected behavior**
An anonymous user viewing the /apis page would not see any results.

**Additional context**
Instead, all API Docs are displayed. When clicking on one of them, you get a proper access denied 403. Seems like the integration between the custom entity and views is not quite complete.

**Comment by shadcn** [comment]
@kscheirer Based on the "Access restrictions" level set on the view. Views will take override any other access set. I confirmed the above steps on a vanilla install.

Anonymous with no permissions to view published content can still access a view with content if "Unrestricted" is set under "Access restrictions".

Hope this helps.

**Comment by kscheirer** [comment]
Thanks for looking into this @arshad!

Same issue exists when view is set with Access: `Permission | View published content`.

So at this point, the anonymous user has `view published content` and *does not have* `view published api docs`

When the anonymous user goes to /apis I wouldnt expect to see any results, instead I see all the API Docs. When the anonymous user goes to view a particular api like api/5, they properly receive a 403.

I was under the impression that Views will not show me lists of content which I can't access, is that not true?

**Comment by shadcn** [comment]
@kscheirer I mean I ran my tests above on a vanilla Drupal and tested the "View published content" permission and this is how it works as well. On kickstarter, we'll need to use "view published api docs" permission on the view but this still shows the restricted APIs (behaves the same way as any other entity in Drupal)

**Comment by kscheirer** [comment]
> I was under the impression that Views will not show me lists of content which I can't access, is that not true?

After some more research, that assumption is totally not true _for custom entities_. For nodes this is still true. Closing this issue, I'll look for another way to solve my use case.

**Comment by micovery** [comment]
Hey Karl @kscheirer , were you able to find a workaround for this. I ran into this as well. I was looking to restrict access to specific APIs both on the /apis page as well as the Cards that are shown in the homepage under the "Featured APIs" Card Group.

**Comment by kscheirer** [comment]
I ended up talking around this during a demo, and did not implement an actual solution. We had 2 ideas though:
1. Create a taxonomy that's used for access control - use it on both api catalog entities and users. Then in the view, add a relationship to the current user, and an argument for the taxonomy you created. This should have the effect that a user will only see api catalog items that are tagged with the same taxonomy terms as their user account. Use a field permission or form_alter to prevent the user field from being edited by non-admins.
2. https://github.com/apigee/apigee-api-catalog-drupal/pull/26 makes api catalog items follow api product access control. Combine with https://gorannikolovski.com/blog/query-level-filtering-custom-entities-drupal-8 to extend that access check to a views query. This also requires Entity API module, the relevant code that was added about a year ago is here: https://www.drupal.org/project/entity/issues/2909970

Sorry I dont have something simpler yet!
